component/loki/source/aws_firehose/internal/handler.go

```diff
@@ -0,0 +1,255 @@
+)
+
+// RecordOrigin is a type that tells from which origin the data received from AWS Firehose comes.
+type RecordOrigin string
+
+const (
+	OriginCloudwatchLogs RecordOrigin = "cloudwatch-logs"
+	OriginDirectPUT      RecordOrigin = "direct-put"
+	OriginUnknown        RecordOrigin = "unknown"
+)
+
+// Sender is an interface that decouples the Firehose request handler from the destination where read loki entries
+// should be written to.
+type Sender interface {
+	Send(ctx context.Context, entry loki.Entry)
+}
+
+// Handler implements a http.Handler that is able to receive records from a Firehose HTTP destination.
+type Handler struct {
+	metrics      *metrics
+	logger       log.Logger
+	sender       Sender
+	relabelRules []*relabel.Config
+}
+
+// NewHandler creates a new handler.
+func NewHandler(sender Sender, logger log.Logger, reg prometheus.Registerer, rbs []*relabel.Config) *Handler {
+	return &Handler{
+		metrics:      newMetrics(reg),
+		logger:       logger,
+		sender:       sender,
+		relabelRules: rbs,
+	}
+}
+
+// ServeHTTP satisfies the http.Handler interface.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	var err error
+	defer req.Body.Close()
+	level.Info(h.logger).Log("msg", "handling request")
+
+	var bodyReader io.Reader = req.Body
+	// firehose allows the user to configure gzip content-encoding, in that case
+	// decompress in the reader during unmarshalling
+	if req.Header.Get("Content-Encoding") == "gzip" {
+		bodyReader, err = gzip.NewReader(req.Body)
+		if err != nil {
+			h.metrics.errorsAPIRequest.WithLabelValues("pre_read").Inc()
+			level.Error(h.logger).Log("msg", "failed to create gzip reader", "err", err.Error())
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+
+	firehoseReq := FirehoseRequest{}
+	err = json.NewDecoder(bodyReader).Decode(&firehoseReq)
+	if err != nil {
+		h.metrics.errorsAPIRequest.WithLabelValues("read_or_format").Inc()
+		level.Error(h.logger).Log("msg", "failed to unmarshall request", "err", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// common labels contains all record-wide labels
+	commonLabels := labels.NewBuilder(nil)
+	commonLabels.Set("__aws_firehose_request_id", req.Header.Get("X-Amz-Firehose-Request-Id"))
+	commonLabels.Set("__aws_firehose_source_arn", req.Header.Get("X-Amz-Firehose-Source-Arn"))
+
+	// if present, use the tenantID header
+	tenantHeader := req.Header.Get("X-Scope-OrgID")
+	if tenantHeader != "" {
+		commonLabels.Set(lokiClient.ReservedLabelTenantID, tenantHeader)
+	}
+
+	h.metrics.batchSize.WithLabelValues().Observe(float64(len(firehoseReq.Records)))
+
+	for _, rec := range firehoseReq.Records {
+		// cleanup err since it might have failed in the previous iteration
+		err = nil
+
+		decodedRecord, recordType, err := h.decodeRecord(rec.Data)
+		if err != nil {
+			h.metrics.errorsRecord.WithLabelValues(getReason(err)).Inc()
```

Comment:
Do we know the cardinality of reason? It seems it could be high, but in actual usage maybe its minimal?

Reply:
Yes, it's something I've added for easily propagating some error codes out of some inner functions. It has <10 values. You can check all the `errWithReason` usages, and if the error is not that type `getReason` fallsback to `unknown`